### PR TITLE
[Inductor] Support one node creating multiple mutations in scheduler

### PR DESCRIPTION
**Pull request by oulgen**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #112801
* #112752
* __->__ #112547

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

### CODE REVIEW

**Comment by oulgen** [comment]
This solution feels pretty janky but fairly straight forward. I'm not 100% confident it is correct but here's an example graph we generate
model___9

**Comment by jansel** [comment]
Can add a test and show a dependency graph with:
1) Multiple triton kernels mutating the same values
2) Non-triton kernels using the mutated data after the triton kernel runs

I am looking for the dependency edge going from the triton kernel to the user after mutation.

**Comment by oulgen** [comment]
@jansel @Chillee 

Updated my solution to use NopOutput IR node instead. I wasn't a big fan of the fake underscored names.

For test case `test_triton_kernel_dependancies `, it uses two kernels and mutates the end result after:
```
      def call_triton(
          x: torch.Tensor,
          y: torch.Tensor,
      ):
          output = torch.zeros_like(x)
          n_elements = output.numel()
          grid = lambda meta: (triton.cdiv(n_elements, meta["BLOCK_SIZE"]),)
          add_kernel_autotuned[grid](x, y, output, n_elements)
          output2 = torch.zeros_like(output)
          add_kernel_autotuned[grid](output, y, output2, n_elements)
          output3 = torch.add(output2, 1)
          return output3
```
the dependency graph is
two_kernel

For the test case `test_triton_kernel_caching` which does 4 kernels in a loop:
multi_kernel

File: torch/_inductor/ir.py [hunk 1]
```diff
@@ -3832,6 +3832,49 @@ def __init__(self, *, kernel_idx, grid, kernel_args):
...
+
+    def has_side_effects(self):
+        return True
+
+    def has_aliasing(self):
+        return True
+
+    def get_alias_names(self):
+        return [self.inputs[0].get_name()]
+
```
Comment (oulgen) [inline torch/_inductor/ir.py:3878]: it is clear to me whether these two are needed or not

**Comment by Chillee** [comment]
Talking with Oguz about the state of mutation in Inductor, and this is my understanding of the current state.

`MutationLayout` is how we often denote mutation in Inductor, and I think this causes a lot of misunderstanding. In particular, there are a number of checks for `MutationLayout` that I think are not sound. `MutationLayout` is mostly a (thin) wrapper over 3 steps:

1. Set `get_mutation_names()`. This is used in the scheduler to introduce mutation dependencies.
2. Call `mark_buffer_mutated`. This realizes all TensorBox ops depending on the mutated buffer, and adds to `mutated_buffers` (which I think is ... not needed).
3. In codegen, we check if its `MutatedLayout`, and if so, we skip allocating a buffer.

Crucially, we do *not* rely on `MutatedLayout` being returned. Thus, we often have a pattern like
```python
def inplace_lowering(inp):
    foo = ir.Foo(MutationLayout(inp))
    V.graph.register_buffer(buffer)
    return inp # foo is never returned!
```

We decided that we should get rid of MutationLayout, and instead use `NoneLayout` whenever we're currently using it. Then, we should use something like `NopOutput` in order to set mutation deps appropriately. This will make it clear that that the IR node with the mutation does not need to be used. To be clear, this is how UserDefinedTritonKernel is currently implemented.

@oulgen will do this refactor in a later PR.

**Review by Chillee — APPROVED** [review]
